### PR TITLE
HBW-61 Add support for Camunda

### PR DESCRIPTION
**Pull request by sugarfree1**

### CODE REVIEW

File: hbw/app/models/hbw/activiti/adapter.rb [hunk 1]
```diff
@@ -1,41 +1,16 @@
...
-        d_pattern = pattern.mb_chars.downcase.to_s
-        users.select do |user|
-          user.values.find { |v| v.mb_chars.downcase.include?(d_pattern) }
-        end
-      end
-
       def process_instances(entity_code, entity_class)
         response = api.post(
-          'query/process-instances', variables: [
+            'query/process-instances', variables: [
```
Comment (GideonEXE) [inline hbw/app/models/hbw/activiti/adapter.rb]: We use 2 spaces for 1 lvl

File: hbw/app/models/hbw/activiti/adapter.rb [hunk 2]
```diff
@@ -49,58 +24,12 @@ def submit(user_email, entity_class, task_id, form_data)
...
-      end
-
-      def form(user_email, entity_class, task_id)
-        task = task_for_email_and_task_id(user_email, entity_class, task_id)
-        HBW::Form.with_connection(api) do
-          HBW::Form.fetch(task, entity_class)
-        end
+      def get_variables(user, entity_class, entity_code)
+        {
+            :initiator                    => {value: user.id, type: :string},
```
Comment (GideonEXE) [inline hbw/app/models/hbw/activiti/adapter.rb]: Same here.

File: hbw/app/models/hbw/camunda/adapter.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
+module HBW
+  module Camunda
+    class Adapter < ::HBW::Common::Adapter
+      include HBW::Inject[:api, :config]
+
+      def process_instances(entity_code, entity_class)
+        response = api.post(
+            '/rest/process-instance', variables: [
```
Comment (GideonEXE) [inline hbw/app/models/hbw/camunda/adapter.rb]: Spaces.

File: hbw/app/models/hbw/camunda/adapter.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+        variables = variables.map { |item| [item.delete(:name), item]}.to_h
+
+        response = api.post("/rest/task/#{task_id}/submit-form", variables: variables)
+        response.status == 204
+      end
+
+      def get_variables(user, entity_class, entity_code)
+        [
+            { name: :initiator,      value: user.id,     type: :string },
```
Comment (GideonEXE) [inline hbw/app/models/hbw/camunda/adapter.rb]: spaces

File: hbw/app/models/hbw/common/adapter.rb [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        user = HBW::BPMUser.with_connection(api) do
+          HBW::BPMUser.fetch(user_email)
+        end
+
+        !user.nil?
+      end
+
+      def users_lookup(pattern)
+        d_pattern = pattern.mb_chars.downcase.to_s
+        users.select do |user|
```
Comment (GideonEXE) [inline hbw/app/models/hbw/common/adapter.rb]: Empty line before select.

File: hbw/app/models/hbw/common/adapter.rb [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+      def submit(_, _, _, _)
+        raise NotImplementedError
+      end
+
+      # TODO: How to distinguish between running process instance and done
+      # TODO: Think of suspended process instances
+      def bp_running?(entity_code, entity_class, current_user_identifier)
+        !process_instances(entity_code, entity_class).empty? ||
+            !task_list_response(current_user_identifier, entity_code, entity_class, 1000, true).empty?
```
Comment (GideonEXE) [inline hbw/app/models/hbw/common/adapter.rb]: Spaces

File: hbw/app/models/hbw/form.rb [hunk 7]
```diff
@@ -4,14 +4,24 @@ class Form
...
         process_definition = task.process_definition
-        deployment = ::HBW::Deployment.fetch(process_definition)
-        resource = deployment.resource(task.form_key)
-        if resource.nil?
-          raise ArgumentError.new(I18n.t('activerecord.errors.models.order.attributes.data.form_key_missed_in_task', form_key: task.form_key))
+
+        if activiti?
+          deployment = ::HBW::Deployment.fetch(process_definition)
+          resource = deployment.resource(task.form_key)
+          if resource.nil?
```
Comment (GideonEXE) [inline hbw/app/models/hbw/form.rb]: raise ..... if resource.nil?

File: hbw/app/models/hbw/task.rb [hunk 8]
```diff
@@ -12,31 +12,64 @@ def method_missing(variable_name)
...
-                       includeProcessVariables: true,
-                       processInstanceVariables: [
-                         name: HBW::Widget.config.fetch(entity_class)[:entity_code_key],
-                         operation: operation(entity_code),
-                         value: entity_code
-                       ],
-                       size: size))
+          if activiti?
+            wrap(
+                do_request(:post,
```
Comment (GideonEXE) [inline hbw/app/models/hbw/task.rb]: Spaces

File: hbw/app/models/hbw/task.rb [hunk 9]
```diff
@@ -12,31 +12,64 @@ def method_missing(variable_name)
...
+                           'query/tasks',
+                           assignee: assignee(user, email, for_all_users),
+                           active: true,
+                           includeProcessVariables: true,
+                           processInstanceVariables: [
+                               name: HBW::Widget.config.fetch(entity_class)[:entity_code_key],
```
Comment (GideonEXE) [inline hbw/app/models/hbw/task.rb]: Spaces

File: hbw/app/models/hbw/task.rb [hunk 10]
```diff
@@ -12,31 +12,64 @@ def method_missing(variable_name)
...
       def wrap(tasks)
         definitions = tasks.map.with_object({}) do |task, d|
-          url = task.fetch('processDefinitionUrl')
+          if activiti?
+            url = task.fetch('processDefinitionUrl')
+          else
+            id = task.fetch('processDefinitionId')
+
+            variables = do_request(:get, "/rest/process-instance/#{task.fetch('processInstanceId')}/variables")
+            task.merge!('variables' => variables.map { |k, v| v.merge({'name' => k})})
```
Comment (GideonEXE) [inline hbw/app/models/hbw/task.rb]: Hash-rocket!?

File: hbw/app/models/hbw/form.rb (hunk 7 above)
Comment (sugarfree1) [reply hbw/app/models/hbw/form.rb]: I force everyone to use
```
if condition?
  raise Exception
end
```

File: hbw/app/models/hbw/task.rb (hunk 10 above)
Comment (sugarfree1) [reply hbw/app/models/hbw/task.rb]: Yes.
```
2.5.1 :001 > {'test': 1}
 => {:test=>1}
2.5.1 :002 > {'test' => 1}
 => {"test"=>1}
```

**Comment by sugarfree1** [comment]
Updated

**Comment by sugarfree1** [comment]
Updated

File: hbw/app/models/hbw/task.rb [hunk 11]
```diff
@@ -12,31 +12,64 @@ def method_missing(variable_name)
...
```
Comment (GideonEXE) [inline hbw/app/models/hbw/task.rb]: `variables.map { |k, v| v.merge({ 'name' => k }) }`

**Comment by sugarfree1** [comment]
Updated

**Comment by sugarfree1** [comment]
Anything else?

**Pull request merged**